gobblin-utility/src/main/java/gobblin/util/filesystem/FileStatusEntry.java

```diff
@@ -0,0 +1,168 @@
+package gobblin.util.filesystem;
+
+import com.google.common.base.Optional;
+import java.io.IOException;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+
+public class FileStatusEntry extends FileStatus {
+
+  static final FileStatusEntry[] EMPTY_ENTRIES = new FileStatusEntry[0];
+
+  private final FileStatusEntry parent;
+  private FileStatusEntry[] children;
+
+  private boolean exists;
+
+  public Optional<FileStatus> _fileStatus;
+
+  public FileStatusEntry(final Path path)
+      throws IOException {
+    this(null, path);
+  }
+
+  public FileStatusEntry(final FileStatusEntry parent, final Path path)
+      throws IOException {
+    if (path == null) {
+      throw new IllegalArgumentException("Path is missing");
+    }
+    this.parent = parent;
+    this._fileStatus = Optional.of(path.getFileSystem(new Configuration()).getFileStatus(path));
+  }
+
+  public boolean refresh(final Path path)
+      throws IOException {
+
+    // cache original values
+    final boolean origExists = exists;
+    if (_fileStatus.isPresent()) {
+      FileStatus oldStatus = _fileStatus.get();
+      Configuration conf = new Configuration();
+
+      try (FileSystem fileSystem = path.getFileSystem(conf)) {
+
+        // refresh the values
+        exists = fileSystem.exists(path);
+        this._fileStatus = Optional.of(fileSystem.getFileStatus(path));
```

Comment:
A follow-up question, do we always favor  `Optional.fromNullable` over ` Optional.of` ?